### PR TITLE
More synchronization optimizations

### PR DESCRIPTION
**Pull request by YunHsiao**
Hi Arseny,

Glad to see the new niagara stream (Your streams are literally my starting point to learn Vulkan back then!), sorry about not fully fixing the problem last time (just touched where the SyncVal then has taken me). I just couldn't stop smiling knowing that you too are struggling with it a bit ;P

So after watching the playback, this time I took a little closer look and found some further improvements over the usages on top of the latest update, and everything still works after the change with no validation error on all paths. Let me know if it helps :)

### CODE REVIEW

File: src/niagara.cpp [hunk 1]
```diff
@@ -789,10 +789,13 @@ int main(int argc, const char** argv)
...
-			wfi.srcAccessMask = VK_ACCESS_MEMORY_WRITE_BIT;
-			wfi.dstAccessMask = VK_ACCESS_MEMORY_READ_BIT;
-			vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, 0, 1, &wfi, 0, 0, 0, 0);
+			VkMemoryBarrier2 barrier = { VK_STRUCTURE_TYPE_MEMORY_BARRIER_2 };
+			barrier.srcStageMask = barrier.dstStageMask = VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
+			barrier.srcAccessMask = barrier.dstAccessMask = VK_ACCESS_MEMORY_READ_BIT | VK_ACCESS_MEMORY_WRITE_BIT;
+			VkDependencyInfo dependencyInfo = { VK_STRUCTURE_TYPE_DEPENDENCY_INFO };
+			dependencyInfo.memoryBarrierCount = 1;
+			dependencyInfo.pMemoryBarriers = &barrier;
+			vkCmdPipelineBarrier2(commandBuffer, &dependencyInfo);
```
Comment (YunHsiao) [inline src/niagara.cpp:798]: Full barriers should specify both read & write for sr & dst access, as per [Synchronization Wiki](https://github.com/KhronosGroup/Vulkan-Docs/wiki/Synchronization-Examples#full-pipeline-barrier)

File: src/niagara.cpp [hunk 2]
```diff
@@ -890,15 +896,14 @@ int main(int argc, const char** argv)
...
 				bufferBarrier(dccb.buffer,
 					VK_PIPELINE_STAGE_TRANSFER_BIT, VK_ACCESS_TRANSFER_WRITE_BIT,
 					VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT),
+				bufferBarrier(dcb.buffer,
+					VK_PIPELINE_STAGE_DRAW_INDIRECT_BIT | rasterizationStage, VK_ACCESS_INDIRECT_COMMAND_READ_BIT | VK_ACCESS_SHADER_READ_BIT,
+					VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, VK_ACCESS_SHADER_WRITE_BIT),
 			};
-
-			pipelineBarrier(commandBuffer, 0, COUNTOF(fillBarriers), fillBarriers, 1, &pyramidBarrier);
+			pipelineBarrier(commandBuffer, 0, late ? COUNTOF(fillBarriers) : 1, fillBarriers, 1, &pyramidBarrier);
```
Comment (YunHsiao) [inline src/niagara.cpp:906]: We wait for device idle every frame so I think for the early cull pass the barrier is redundant.

File: src/niagara.cpp [hunk 3]
```diff
@@ -1058,10 +1064,10 @@ int main(int argc, const char** argv)
 		VkImageMemoryBarrier2 renderBeginBarriers[] =
 		{
 			imageBarrier(colorTarget.image,
-				VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT, 0, VK_IMAGE_LAYOUT_UNDEFINED,
+				0, 0, VK_IMAGE_LAYOUT_UNDEFINED,
```
Comment (YunHsiao) [inline src/niagara.cpp:1067]: As per [spec](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkPipelineStageFlagBits2.html):
> The TOP and BOTTOM pipeline stages are deprecated, and applications should prefer VK_PIPELINE_STAGE_2_ALL_COMMANDS_BIT and VK_PIPELINE_STAGE_2_NONE.

And we don't actually need to wait for anything at the beginning of the frame here, so `VK_PIPELINE_STAGE_2_NONE` is suffice.

File: src/niagara.cpp [hunk 4]
```diff
@@ -1143,7 +1149,7 @@ int main(int argc, const char** argv)
 
 		VkImageMemoryBarrier2 presentBarrier = imageBarrier(swapchain.images[imageIndex],
 			VK_PIPELINE_STAGE_TRANSFER_BIT, VK_ACCESS_TRANSFER_WRITE_BIT, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL,
-			VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 0, VK_IMAGE_LAYOUT_PRESENT_SRC_KHR);
+			0, 0, VK_IMAGE_LAYOUT_PRESENT_SRC_KHR);
```
Comment (YunHsiao) [inline src/niagara.cpp:1152]: And block nothing here, before present

File: src/niagara.cpp [hunk 5]
```diff
@@ -1040,7 +1045,8 @@ int main(int argc, const char** argv)
 
 				VkImageMemoryBarrier2 reduceBarrier = imageBarrier(depthPyramid.image,
 					VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, VK_ACCESS_SHADER_WRITE_BIT, VK_IMAGE_LAYOUT_GENERAL,
-					VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, VK_ACCESS_SHADER_READ_BIT, VK_IMAGE_LAYOUT_GENERAL);
+					VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, VK_ACCESS_SHADER_READ_BIT, VK_IMAGE_LAYOUT_GENERAL,
+					VK_IMAGE_ASPECT_COLOR_BIT, i, 1);
```
Comment (YunHsiao) [inline src/niagara.cpp:1049]: I think this is another bug in the validation layer, we need to do per-level transition here but it didn't report any error before this change...

File: src/niagara.cpp (hunk 2 above)
Comment (zeux) [reply src/niagara.cpp:906]: fwiw I wouldn't want to rely on this long term, as it's a pretty obvious shortcut that we just haven't had a chance to fix.

File: src/niagara.cpp (hunk 1 above)
Comment (zeux) [reply src/niagara.cpp:798]: Ah yeah that makes sense, our previous barrier had WAR hazards.

File: src/niagara.cpp (hunk 2 above)
Comment (zeux) [reply src/niagara.cpp:906]: This also makes it a little more awkward to change things in the future, so I'd keep this one as is.

Comment (YunHsiao) [reply src/niagara.cpp]: All right, for future maneuvers, kept this among some others as is.

File: src/niagara.cpp (hunk 4 above)
Comment (YunHsiao) [reply src/niagara.cpp:1152]: Present barriers still don't need to block anything, since it already waits for the submit semaphore.

**Comment by zeux** [comment]
Thanks!

**Pull request merged**